# **ENTREGABLE 1**
## Kevin Andre Baracaldo

In [1]:
# Importar las librerias necesarias

import pandas as pd
import requests
import sqlite3

In [2]:
# ===================================================================
# PASO 1: CONFIGURACIÓN E INSTALACIÓN DE LIBRERÍAS
# ===================================================================

import pandas as pd
import requests
import sqlite3

print("✅ Librerías importadas.")

# ===================================================================
# PASO 2: EXTRACCIÓN (EXTRACT) DE LOS DATOS
# ===================================================================

# La plataforma datos.gov.co usa la API de Socrata. Podemos usarla para
# descargar los datos directamente, lo que es más eficiente que bajar un CSV.
# Aumentamos el límite para traer más filas (ajusta si es necesario).
api_url = "https://www.datos.gov.co/resource/nudc-7mev.json?$limit=50000"

print(f"📥 Extrayendo datos desde: {api_url}")

try:
    response = requests.get(api_url)
    response.raise_for_status()  # Lanza un error si la petición falla (ej: 404)
    data = response.json()
    df_raw = pd.DataFrame(data)
    print(f"✅ ¡Extracción exitosa! Se cargaron {len(df_raw)} filas.")
    display(df_raw.head())

except requests.exceptions.RequestException as e:
    print(f"❌ Error al extraer los datos: {e}")
    df_raw = pd.DataFrame() # Creamos un dataframe vacío para evitar errores posteriores

except Exception as e:
    print(f"❌ Ocurrió un error inesperado: {e}")
    df_raw = pd.DataFrame()

✅ Librerías importadas.
📥 Extrayendo datos desde: https://www.datos.gov.co/resource/nudc-7mev.json?$limit=50000
✅ ¡Extracción exitosa! Se cargaron 14585 filas.


a_o c_digo_municipio   municipio c_digo_departamento departamento  \
0  2023            05004    Abriaquí                  05    Antioquia   
1  2023            95025  El Retorno                  95     Guaviare   
2  2023            95200  Miraflores                  95     Guaviare   
3  2023            97001        Mitú                  97       Vaupés   
4  2023            97161      Caruru                  97       Vaupés   

  c_digo_etc              etc poblaci_n_5_16 tasa_matriculaci_n_5_16  \
0       3758  Antioquia (ETC)            503                   62.62   
1       3830   Guaviare (ETC)           4438                   53.27   
2       3830   Guaviare (ETC)           2014                   32.52   
3       3831     Vaupés (ETC)          10986                   59.57   
4       3831     Vaupés (ETC)           1228                    51.3   

  cobertura_neta  ... reprobaci_n_primaria reprobaci_n_secundaria  \
0          62.62  ...                 1.96                  16.51   
1          53.27  ...                 7.11                   9.39   
2          32.52  ...                 6.93                  14.13   
3          59.57  ...                 4.04                   8.33   
4           51.3  ...                 7.32                  15.28   

  reprobaci_n_media repitencia repitencia_transici_n repitencia_primaria  \
0              2.04       9.52                     0               10.46   
1              1.75       9.34                  6.95               11.84   
2              7.81       8.65                  6.67                9.04   
3               4.6      16.18                  7.75               21.04   
4              7.27       9.24                  2.86                7.62   

  repitencia_secundaria repitencia_media tama_o_promedio_de_grupo  \
0                 13.76             2.04                      NaN   
1                  8.48             3.16                      NaN   
2                 10.25             1.54                      NaN   
3                 13.84             7.18                      NaN   
4                 14.85             3.64                      NaN   

  sedes_conectadas_a_internet  
0                         NaN  
1                         NaN  
2                         NaN  
3                         NaN  
4                         NaN  

[5 rows x 41 columns]

In [ ]:
display(df_raw)

a_o c_digo_municipio    municipio c_digo_departamento departamento  \
0      2023            05004     Abriaquí                  05    Antioquia   
1      2023            95025   El Retorno                  95     Guaviare   
2      2023            95200   Miraflores                  95     Guaviare   
3      2023            97001         Mitú                  97       Vaupés   
4      2023            97161       Caruru                  97       Vaupés   
...     ...              ...          ...                 ...          ...   
14580  2011             5036  Angelópolis                   5    Antioquia   
14581  2011             5034        Andes                   5    Antioquia   
14582  2011             5031       Amalfi                   5    Antioquia   
14583  2011             5030        Amagá                   5    Antioquia   
14584  2011             5021   Alejandría                   5    Antioquia   

      c_digo_etc              etc poblaci_n_5_16 tasa_matriculaci_n_5_16  \
0           3758  Antioquia (ETC)            503                   62.62   
1           3830   Guaviare (ETC)           4438                   53.27   
2           3830   Guaviare (ETC)           2014                   32.52   
3           3831     Vaupés (ETC)          10986                   59.57   
4           3831     Vaupés (ETC)           1228                    51.3   
...          ...              ...            ...                     ...   
14580       3758  Antioquia (ETC)           1707                   78.85   
14581       3758  Antioquia (ETC)          10244                   84.45   
14582       3758  Antioquia (ETC)           5552                   97.71   
14583       3758  Antioquia (ETC)           6631                   78.65   
14584       3758  Antioquia (ETC)            702                  118.52   

      cobertura_neta  ... reprobaci_n_primaria reprobaci_n_secundaria  \
0              62.62  ...                 1.96                  16.51   
1              53.27  ...                 7.11                   9.39   
2              32.52  ...                 6.93                  14.13   
3              59.57  ...                 4.04                   8.33   
4               51.3  ...                 7.32                  15.28   
...              ...  ...                  ...                    ...   
14580           78.9  ...                 3.61                    9.5   
14581           84.5  ...                 0.58                   0.04   
14582           97.7  ...                    0                      0   
14583           78.7  ...                 6.73                  14.46   
14584          118.5  ...                 2.16                   4.39   

      reprobaci_n_media repitencia repitencia_transici_n repitencia_primaria  \
0                  2.04       9.52                     0               10.46   
1                  1.75       9.34                  6.95               11.84   
2                  7.81       8.65                  6.67                9.04   
3                   4.6      16.18                  7.75               21.04   
4                  7.27       9.24                  2.86                7.62   
...                 ...        ...                   ...                 ...   
14580              7.32       0.71                     0                 0.7   
14581              2.69       5.41                  0.73                5.53   
14582                 0        NaN                  0.83                 NaN   
14583              7.45       0.42                     0                0.24   
14584              8.04        0.6                     0                0.96   

      repitencia_secundaria repitencia_media tama_o_promedio_de_grupo  \
0                     13.76             2.04                      NaN   
1                      8.48             3.16                      NaN   
2                     10.25             1.54                      NaN   
3                     13.84 

La base de datos reúne 14.585 observaciones distribuidas en 41 variables que reflejan diversas dimensiones del sistema educativo colombiano en los niveles de preescolar, básica y media, clasificadas por municipio y año. Se registran indicadores clave de cobertura educativa, como la tasa de matrícula y los porcentajes de cobertura neta y bruta; así como métricas de desempeño del sistema, entre ellas los índices de aprobación, reprobación, repitencia y deserción. Adicionalmente, incorpora datos sobre infraestructura educativa, tales como el promedio de estudiantes por grupo y la proporción de sedes con acceso a internet. Las variables territoriales permiten segmentar los datos por regiones. Cabe resaltar que existen vacíos de información en algunos campos, principalmente en los relacionados con infraestructura, lo cual requerirá un tratamiento específico durante el análisis.

In [4]:
# ===================================================================
# PASO 3: EXTRACCIÓN BASE DIVIPOLA
# ===================================================================

print("📥 Extrayendo datos de DIVIPOLA desde su API...")

api_url_divipola = "https://www.datos.gov.co/resource/gdxc-w37w.json?$limit=2000"

try:
    response = requests.get(api_url_divipola)
    response.raise_for_status()
    data = response.json()
    df_divipola = pd.DataFrame(data)
    print(f"✅ ¡Base DIVIPOLA cargada! {len(df_divipola)} filas.")
except requests.exceptions.RequestException as e:
    print(f"❌ Error al extraer DIVIPOLA: {e}")
    df_divipola = pd.DataFrame()
except Exception as e:
    print(f"❌ Error inesperado: {e}")
    df_divipola = pd.DataFrame()


📥 Extrayendo datos de DIVIPOLA desde su API...


✅ ¡Base DIVIPOLA cargada! 1122 filas.


In [5]:
df_divipola

cod_dpto       dpto cod_mpio        nom_mpio          tipo_municipio  \
0          05  ANTIOQUIA    05001        MEDELLÍN               Municipio   
1          05  ANTIOQUIA    05002       ABEJORRAL               Municipio   
2          05  ANTIOQUIA    05004        ABRIAQUÍ               Municipio   
3          05  ANTIOQUIA    05021      ALEJANDRÍA               Municipio   
4          05  ANTIOQUIA    05030           AMAGÁ               Municipio   
...       ...        ...      ...             ...                     ...   
1117       97     VAUPÉS    97889        YAVARATÉ  Área no municipalizada   
1118       99    VICHADA    99001  PUERTO CARREÑO               Municipio   
1119       99    VICHADA    99524    LA PRIMAVERA               Municipio   
1120       99    VICHADA    99624   SANTA ROSALÍA               Municipio   
1121       99    VICHADA    99773        CUMARIBO               Municipio   

        longitud   latitud  
0     -75,581775  6,246631  
1     -75,428739  5,789315  
2     -76,064304  6,632282  
3     -75,141346  6,376061  
4     -75,702188  6,038708  
...          ...       ...  
1117  -69,203337  0,609142  
1118  -67,487095  6,186636  
1119  -70,410515  5,486309  
1120  -70,859499  5,136393  
1121  -69,795533  4,446352  

[1122 rows x 7 columns]

La base DIVIPOLA (División Político-Administrativa de Colombia) es la fuente oficial del DANE encargada de codificar y organizar los distintos niveles territoriales del país. Proporciona códigos y nombres estandarizados para departamentos, municipios, corregimientos, centros poblados y cabeceras municipales, facilitando la homogeneización geográfica en conjuntos de datos gubernamentales. Cada municipio cuenta con un código único de seis cifras: los dos primeros dígitos corresponden al departamento y los cuatro siguientes al municipio.

Esta base será empleada para construir una de las dimensiones de nuestro modelo de datos.

## Limpieza de datos

In [6]:
pd.DataFrame(df_raw.columns.tolist(), columns=["Columnas originales"])


Columnas originales
0                           a_o
1              c_digo_municipio
2                     municipio
3           c_digo_departamento
4                  departamento
5                    c_digo_etc
6                           etc
7                poblaci_n_5_16
8       tasa_matriculaci_n_5_16
9                cobertura_neta
10    cobertura_neta_transici_n
11      cobertura_neta_primaria
12    cobertura_neta_secundaria
13         cobertura_neta_media
14              cobertura_bruta
15   cobertura_bruta_transici_n
16     cobertura_bruta_primaria
17   cobertura_bruta_secundaria
18        cobertura_bruta_media
19                    deserci_n
20         deserci_n_transici_n
21           deserci_n_primaria
22         deserci_n_secundaria
23              deserci_n_media
24                   aprobaci_n
25        aprobaci_n_transici_n
26          aprobaci_n_primaria
27        aprobaci_n_secundaria
28             aprobaci_n_media
29                  reprobaci_n
30       reprobaci_n_transici_n
31         reprobaci_n_primaria
32       reprobaci_n_secundaria
33            reprobaci_n_media
34                   repitencia
35        repitencia_transici_n
36          repitencia_primaria
37        repitencia_secundaria
38             repitencia_media
39     tama_o_promedio_de_grupo
40  sedes_conectadas_a_internet

In [15]:
!pip install unidecode


In [16]:
import unidecode

# Copia de trabajo
df = df_raw.copy()

# 1. Eliminar columnas completamente vacías
df = df.dropna(axis=1, how='all')

# 2. Renombrar columnas: quitar tildes, espacios, mayúsculas
df.columns = (
    df.columns
    .str.strip()
    .str.lower()
    .str.replace(' ', '_')
    .str.replace(r'[^a-z0-9_]', '', regex=True)
    .map(unidecode.unidecode)  # quitar tildes
)

# 3. Eliminar filas del departamento 'NACIONAL' o similares
df = df[~df['departamento'].str.upper().str.strip().isin(['NACIONAL'])]

# 4. Normalizar texto de municipio y departamento
df['municipio'] = df['municipio'].astype(str).str.strip().str.upper().map(unidecode.unidecode)
df['departamento'] = df['departamento'].astype(str).str.strip().str.upper().map(unidecode.unidecode)

# 5. Convertir columnas numéricas (las que se puedan)
for col in df.columns:
    df[col] = pd.to_numeric(df[col], errors='ignore')

# 6. Asegurar que la columna de año esté como entero
if 'ano' in df.columns:
    df['ano'] = pd.to_numeric(df['ano'], errors='coerce').astype('Int64')

# 7. Resetear índice
df.reset_index(drop=True, inplace=True)

# Guardar en session_state o variable limpia
df_clean = df.copy()
print("✅ Transformación completa aplicada correctamente.")


C:\Users\kevin\AppData\Local\Temp\ipykernel_10248\2680745322.py:28: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead
  df[col] = pd.to_numeric(df[col], errors='ignore')


✅ Transformación completa aplicada correctamente.


In [17]:
df_clean

a_o  c_digo_municipio    municipio  c_digo_departamento departamento  \
0      2023              5004     ABRIAQUI                    5    ANTIOQUIA   
1      2023             95025   EL RETORNO                   95     GUAVIARE   
2      2023             95200   MIRAFLORES                   95     GUAVIARE   
3      2023             97001         MITU                   97       VAUPES   
4      2023             97161       CARURU                   97       VAUPES   
...     ...               ...          ...                  ...          ...   
14577  2011              5036  ANGELOPOLIS                    5    ANTIOQUIA   
14578  2011              5034        ANDES                    5    ANTIOQUIA   
14579  2011              5031       AMALFI                    5    ANTIOQUIA   
14580  2011              5030        AMAGA                    5    ANTIOQUIA   
14581  2011              5021   ALEJANDRIA                    5    ANTIOQUIA   

       c_digo_etc              etc poblaci_n_5_16  tasa_matriculaci_n_5_16  \
0          3758.0  Antioquia (ETC)            503                    62.62   
1          3830.0   Guaviare (ETC)           4438                    53.27   
2          3830.0   Guaviare (ETC)           2014                    32.52   
3          3831.0     Vaupés (ETC)          10986                    59.57   
4          3831.0     Vaupés (ETC)           1228                    51.30   
...           ...              ...            ...                      ...   
14577      3758.0  Antioquia (ETC)           1707                    78.85   
14578      3758.0  Antioquia (ETC)          10244                    84.45   
14579      3758.0  Antioquia (ETC)           5552                    97.71   
14580      3758.0  Antioquia (ETC)           6631                    78.65   
14581      3758.0  Antioquia (ETC)            702                   118.52   

       cobertura_neta  ...  reprobaci_n_primaria  reprobaci_n_secundaria  \
0               62.62  ...                  1.96                   16.51   
1               53.27  ...                  7.11                    9.39   
2               32.52  ...                  6.93                   14.13   
3               59.57  ...                  4.04                    8.33   
4               51.30  ...                  7.32                   15.28   
...               ...  ...                   ...                     ...   
14577           78.90  ...                  3.61                    9.50   
14578           84.50  ...                  0.58                    0.04   
14579           97.70  ...                  0.00                    0.00   
14580           78.70  ...                  6.73                   14.46   
14581          118.50  ...                  2.16                    4.39   

       reprobaci_n_media  repitencia  repitencia_transici_n  \
0                   2.04        9.52                   0.00   
1                   1.75        9.34                   6.95   
2                   7.81        8.65                   6.67   
3                   4.60       16.18                   7.75   
4                   7.27        9.24                   2.86   
...                  ...         ...                    ...   
14577               7.32        0.71                   0.00   
14578               2.69        5.41                   0.73   
14579               0.00         NaN                   0.83   
14580               7.45        0.42                   0.00   
14581               8.04        0.60                   0.00   

       repitencia_primaria  repitencia_secundaria  repitencia_media  \
0                    10.46                  13.76              2.04   
1                    11.84                   8.48              3.16   
2                     9.04                  10.25              1.54   
3                    21.04                  13.84              7.18   
4                     7.62                  14.85              3.64   
...           

In [22]:
# Copia de trabajo
df_geo = df_divipola.copy()

# 1. Normalizar columnas
df_geo.columns = (
    df_geo.columns
    .str.strip()
    .str.lower()
    .str.replace(' ', '_')
    .str.replace(r'[^a-z0-9_]', '', regex=True)
)

# 2. Limpiar y estandarizar nombres
df_geo['dpto'] = (
    df_geo['dpto']
    .astype(str)
    .str.strip()
    .str.upper()
    .str.normalize('NFKD')
    .str.encode('ascii', errors='ignore')
    .str.decode('utf-8')
)

df_geo['nom_mpio'] = (
    df_geo['nom_mpio']
    .astype(str)
    .str.strip()
    .str.upper()
    .str.normalize('NFKD')
    .str.encode('ascii', errors='ignore')
    .str.decode('utf-8')
)

# 3. Rellenar ceros en códigos
df_geo['cod_dpto'] = df_geo['cod_dpto'].astype(str).str.zfill(2)
df_geo['cod_mpio'] = df_geo['cod_mpio'].astype(str).str.zfill(3)

# 4. Crear ID geográfico único (6 dígitos)
df_geo['id_geo'] = df_geo['cod_dpto'] + df_geo['cod_mpio']

# 5. Eliminar duplicados
df_geo = df_geo.drop_duplicates(subset='id_geo')

# 6. Resetear índice
df_geo.reset_index(drop=True, inplace=True)

# 7. Renombrar columnas clave para modelo dimensional
df_geo = df_geo.rename(columns={
    'dpto': 'departamento',
    'nom_mpio': 'municipio'
})

print("✅ Base DIVIPOLA transformada exitosamente y lista para usar como dimensión geográfica.")


✅ Base DIVIPOLA transformada exitosamente y lista para usar como dimensión geográfica.


In [23]:
import os
import sqlite3


# Crear conexión SQLite (archivo .db en la carpeta Datos)
conn = sqlite3.connect("../Datos/educacion.db")

# Guardar DataFrames como tablas SQL
df_clean.to_sql("educacion_raw", conn, if_exists="replace", index=False)
df_geo.to_sql("divipola", conn, if_exists="replace", index=False)

print("📁 Tablas 'educacion_raw' y 'divipola' guardadas en SQLite.")


📁 Tablas 'educacion_raw' y 'divipola' guardadas en SQLite.


## Dimensiones para el modelo estrella

In [32]:
# Eliminar si ya existe
conn.execute("DROP TABLE IF EXISTS dim_tiempo;")

# Crear la dimensión de tiempo
query_dim_tiempo = """
CREATE TABLE dim_tiempo AS
SELECT 
    ROW_NUMBER() OVER () AS id_tiempo,
    a_o AS anio
FROM educacion_raw
WHERE a_o IS NOT NULL
GROUP BY a_o
ORDER BY a_o;
"""

conn.execute(query_dim_tiempo)


In [29]:
# Eliminar si ya existe
conn.execute("DROP TABLE IF EXISTS dim_municipio;")

# Crear la dimensión con los nombres exactos de columnas
query_dim_municipio = """
CREATE TABLE dim_municipio AS
SELECT 
    ROW_NUMBER() OVER () AS id_municipio,
    d.id_geo,
    d.cod_dpto AS codigo_departamento,
    d.departamento,
    d.cod_mpio AS codigo_municipio,
    d.municipio,
    d.tipo_municipio,
    d.latitud,
    d.longitud
FROM divipola d
GROUP BY 
    d.id_geo,
    d.cod_dpto,
    d.departamento,
    d.cod_mpio,
    d.municipio,
    d.tipo_municipio,
    d.latitud,
    d.longitud;
"""

conn.execute(query_dim_municipio)


In [36]:
# Eliminar si ya existe
conn.execute("DROP TABLE IF EXISTS dim_departamento;")

# Crear la tabla con el nombre original exacto de la columna
query_dim_departamento = """
CREATE TABLE dim_departamento AS
SELECT 
    ROW_NUMBER() OVER () AS id_departamento,
    c_digo_departamento AS codigo_departamento,
    departamento
FROM educacion_raw
WHERE c_digo_departamento IS NOT NULL
GROUP BY c_digo_departamento, departamento;
"""

conn.execute(query_dim_departamento)


## Tabla de hechos

In [39]:
# Eliminar si ya existe
conn.execute("DROP TABLE IF EXISTS fact_educacion;")

# Crear nueva tabla de hechos con métricas clave
query_fact_educacion = """
CREATE TABLE fact_educacion AS
SELECT 
    t.id_tiempo,
    m.id_municipio,
    d.id_departamento,
    e.tasa_matriculaci_n_5_16 AS tasa_matriculacion,
    e.cobertura_neta,
    e.aprobaci_n AS aprobacion,
    e.reprobaci_n AS reprobacion,
    e.repitencia,
    e.deserci_n AS desercion
FROM educacion_raw e
JOIN dim_tiempo t 
    ON e.a_o = t.anio
JOIN dim_municipio m 
    ON CAST(e.c_digo_municipio AS TEXT) = m.codigo_municipio
JOIN dim_departamento d 
    ON CAST(e.c_digo_departamento AS TEXT) = d.codigo_departamento
WHERE 
    e.tasa_matriculaci_n_5_16 IS NOT NULL
    AND e.cobertura_neta IS NOT NULL;
"""

conn.execute(query_fact_educacion)


# Preguntas

## 1. Respecto a la población del municipio ¿Que porcentaje de escolaridad hay?

In [41]:
query_top_municipios = """
SELECT 
    m.municipio,
    ROUND(AVG(h.tasa_matriculacion), 2) AS tasa_promedio
FROM fact_educacion h
JOIN dim_municipio m 
    ON h.id_municipio = m.id_municipio
GROUP BY m.municipio
HAVING tasa_promedio IS NOT NULL
ORDER BY tasa_promedio DESC
LIMIT 10;
"""

df_top_municipios = pd.read_sql(query_top_municipios, conn)
df_top_municipios


municipio  tasa_promedio
0                 COTA         209.95
1                TENJO         149.60
2            LA CALERA         144.08
3              SALENTO         139.27
4     SABANA DE TORRES         132.08
5        PUERTO GAITAN         129.72
6              FUQUENE         128.59
7           SUBACHOQUE         127.24
8      PUENTE NACIONAL         125.73
9  LA JAGUA DE IBIRICO         122.64

### 📊 Interpretación de los resultados

El análisis de la tasa promedio de matriculación por municipio revela diferencias significativas en la cobertura educativa a nivel local. Municipios como **Cota**, **Tenjo** y **La Calera** presentan los valores más altos, lo cual sugiere un amplio acceso a la educación en comparación con otros territorios. Estos resultados podrían estar relacionados con una buena infraestructura escolar, una mayor oferta educativa o condiciones demográficas particulares.

En particular, el caso de **Cota**, con una tasa promedio de matriculación del **209.95%**, indica que el número de estudiantes matriculados es más del doble de la población estimada entre 5 y 16 años. Este tipo de valores, aunque poco comunes, pueden explicarse por:

- La inclusión de estudiantes fuera del rango de edad oficial (por rezago o adelantamiento),
- La recepción de estudiantes provenientes de otros municipios (migración escolar),
- O posibles diferencias entre la población proyectada y la atendida realmente.

Tasas superiores al 100% no necesariamente reflejan un error, pero sí invitan a profundizar en el contexto educativo y demográfico de cada territorio.


## 2. ¿Cómo compararía el rendimiento educativo por municipios?

In [42]:
query_rendimiento_municipal = """
SELECT 
    m.municipio,
    ROUND(AVG(f.aprobacion), 2) AS promedio_aprobacion,
    ROUND(AVG(f.reprobacion), 2) AS promedio_reprobacion,
    ROUND(AVG(f.repitencia), 2) AS promedio_repitencia,
    ROUND(AVG(f.desercion), 2) AS promedio_desercion
FROM fact_educacion f
JOIN dim_municipio m ON f.id_municipio = m.id_municipio
GROUP BY m.municipio
HAVING promedio_aprobacion IS NOT NULL
ORDER BY promedio_aprobacion DESC
LIMIT 10;
"""

df_rendimiento = pd.read_sql(query_rendimiento_municipal, conn)
df_rendimiento


municipio  promedio_aprobacion  promedio_reprobacion  \
0     PUERTO ARICA                98.97                  0.03   
1         TIBIRITA                98.42                  0.43   
2         LA PALMA                98.28                  0.20   
3        GUATAVITA                98.27                  0.27   
4         CHAGUANI                98.19                  0.16   
5         PUPIALES                98.10                  0.67   
6  MIRITI - PARANA                98.06                  0.38   
7        SUTATAUSA                97.96                  0.11   
8           AGUADA                97.72                  0.90   
9            JUNIN                97.68                  0.98   

   promedio_repitencia  promedio_desercion  
0                 7.43                1.31  
1                 0.83                1.16  
2                 1.58                1.52  
3                 2.62                1.45  
4                 3.00                1.65  
5                 2.05                1.23  
6                 5.98                1.70  
7                 1.03                1.93  
8                 1.03                1.39  
9                 2.85                1.34

### 🎓 Municipios con mejor rendimiento educativo

Este análisis muestra los 10 municipios con mayor promedio de aprobación escolar. Todos superan el 97.5% de aprobación, lo que refleja un alto desempeño en términos de permanencia y logro académico dentro del sistema educativo.

Municipios como **Puerto Arica**, **Tibiritá** y **La Palma** encabezan el listado, combinando altos niveles de aprobación con niveles bajos de reprobación y deserción. Sin embargo, es importante observar que algunos casos, como **Puerto Arica** y **Miriti - Paraná**, presentan niveles elevados de repitencia, lo que podría indicar desafíos en la progresión escolar a pesar de que los estudiantes finalmente aprueban.

Este tipo de análisis permite identificar territorios con buenas prácticas educativas, pero también aquellos donde, a pesar de las altas tasas de aprobación, pueden existir patrones de repitencia o abandono que deben ser atendidos para garantizar trayectorias educativas continuas y efectivas.


## 3. ¿Que departamentos son los que mejor cobertura tienen? ¿Pueden hacer cálculo con SQL?

In [44]:
query_cobertura_ponderada_fix = """
SELECT 
    UPPER(REPLACE(REPLACE(d.departamento, ',', ''), '.', '')) AS departamento,
    ROUND(SUM(e.cobertura_neta * e.poblaci_n_5_16) / SUM(e.poblaci_n_5_16), 2) AS cobertura_neta_ponderada
FROM educacion_raw e
JOIN dim_departamento d 
    ON CAST(e.c_digo_departamento AS TEXT) = d.codigo_departamento
WHERE 
    e.cobertura_neta IS NOT NULL 
    AND e.poblaci_n_5_16 IS NOT NULL 
    AND e.poblaci_n_5_16 > 0
GROUP BY UPPER(REPLACE(REPLACE(d.departamento, ',', ''), '.', ''))
ORDER BY cobertura_neta_ponderada DESC
LIMIT 20;
"""

df_resultado = pd.read_sql(query_cobertura_ponderada_fix, conn)
df_resultado


departamento  cobertura_neta_ponderada
0             CASANARE                     96.97
1                SUCRE                     95.38
2            SANTANDER                     94.02
3            MAGDALENA                     93.71
4              BOLIVAR                     93.51
5            RISARALDA                     92.53
6         CUNDINAMARCA                     92.22
7                 META                     92.09
8                CESAR                     91.98
9            ANTIOQUIA                     91.24
10              TOLIMA                     91.22
11           ATLANTICO                     91.03
12             CORDOBA                     90.81
13               HUILA                     90.01
14  NORTE DE SANTANDER                     89.23
15              BOYACA                     89.02
16           BOGOTA DC                     88.63
17             QUINDIO                     88.44
18          LA GUAJIRA                     84.43
19               CAUCA                     83.79

### 🗺️ Departamentos con mejor cobertura educativa (ponderada)

El análisis de cobertura neta ponderada permite identificar los departamentos donde una mayor proporción de la población entre 5 y 16 años está efectivamente matriculada. A diferencia de un promedio simple, esta métrica ajusta la cobertura con base en el tamaño de la población infantil, ofreciendo una visión más justa y representativa.

Los resultados muestran que departamentos como **Casanare (96.97%)**, **Sucre (95.38%)** y **Santander (94.02%)** lideran en cobertura, lo que sugiere una adecuada capacidad del sistema educativo para atender a la población escolar en esos territorios.

En contraste, aunque **Bogotá D.C.** se encuentra en el listado, su cobertura ponderada es del **88.63%**, lo que indica oportunidades de mejora, especialmente considerando su rol como capital y centro educativo nacional.

Este tipo de análisis es útil para priorizar intervenciones y monitorear el desempeño regional en términos de acceso efectivo a la educación.
